### ISSUE TITLE
Parallel downloads thumbnails pictures / Concurrent server requests

### ISSUE THREAD

**Issue by kfirfer** [nextcloud/android#6855]
The android app making calls to the server synchronized when fetching thumbnails, one thumbnail at a time

In the web version its parallel, sending multiple API calls at once

I think it will improve performance significant in the android app if he will be parallel thumbnails fetch's as-well 

App version: 3.13.0
Android version: 10

**Comment by tobiasKaminsky** [comment]
Totally true.
I am not sure what holds it back, but I fear that our underlying http library is somehow preventing this.

**Comment by tobiasKaminsky** [comment]
I started 50 threads with some remote operation, all were started in the same second:
```
1598858210.881 com.nextcloud.client D/HttpClient: 203063136 start
1598858210.886 com.nextcloud.client D/HttpClient: 240626713 start
1598858210.888 com.nextcloud.client D/HttpClient: 220903103 start
1598858210.904 com.nextcloud.client D/HttpClient: 263102698 start
1598858210.904 com.nextcloud.client D/HttpClient: 254669013 start
1598858210.905 com.nextcloud.client D/HttpClient: 22937563 start
1598858210.906 com.nextcloud.client D/HttpClient: 123525752 start
1598858210.906 com.nextcloud.client D/HttpClient: 233424209 start
1598858210.906 com.nextcloud.client D/HttpClient: 145931959 start
1598858210.907 com.nextcloud.client D/HttpClient: 72212918 start
1598858210.907 com.nextcloud.client D/HttpClient: 46212644 start
1598858210.907 com.nextcloud.client D/HttpClient: 162817421 start
1598858210.910 com.nextcloud.client D/HttpClient: 13525826 start
1598858210.910 com.nextcloud.client D/HttpClient: 19566419 start
1598858210.911 com.nextcloud.client D/HttpClient: 66879632 start
1598858210.911 com.nextcloud.client D/HttpClient: 115569038 start
1598858210.912 com.nextcloud.client D/HttpClient: 17747337 start
1598858210.912 com.nextcloud.client D/HttpClient: 222612911 start
1598858210.913 com.nextcloud.client D/HttpClient: 22091196 start
1598858210.913 com.nextcloud.client D/HttpClient: 260946245 start
1598858210.913 com.nextcloud.client D/HttpClient: 237574298 start
1598858210.914 com.nextcloud.client D/HttpClient: 236623307 start
1598858210.915 com.nextcloud.client D/HttpClient: 190162113 start
1598858210.915 com.nextcloud.client D/HttpClient: 166732902 start
1598858210.915 com.nextcloud.client D/HttpClient: 16241064 start
1598858210.916 com.nextcloud.client D/HttpClient: 54182823 start
1598858210.918 com.nextcloud.client D/HttpClient: 30171220 start
1598858210.918 com.nextcloud.client D/HttpClient: 240569341 start
1598858210.920 com.nextcloud.client D/HttpClient: 205910851 start
1598858210.920 com.nextcloud.client D/HttpClient: 45690304 start
1598858210.920 com.nextcloud.client D/HttpClient: 264904946 start
1598858210.920 com.nextcloud.client D/HttpClient: 122235641 start
1598858210.922 com.nextcloud.client D/HttpClient: 181754430 start
1598858210.922 com.nextcloud.client D/HttpClient: 139100319 start
1598858210.924 com.nextcloud.client D/HttpClient: 82011628 start
1598858210.925 com.nextcloud.client D/HttpClient: 241728949 start
1598858210.926 com.nextcloud.client D/HttpClient: 249595835 start
1598858210.927 com.nextcloud.client D/HttpClient: 130604106 start
1598858210.927 com.nextcloud.client D/HttpClient: 77001944 start
1598858210.928 com.nextcloud.client D/HttpClient: 166740017 start
1598858210.929 com.nextcloud.client D/HttpClient: 96498454 start
1598858210.930 com.nextcloud.client D/HttpClient: 83003012 start
1598858210.976 com.nextcloud.client D/HttpClient: 148324719 start
1598858210.978 com.nextcloud.client D/HttpClient: 130069253 start
1598858210.981 com.nextcloud.client D/HttpClient: 12617099 start
1598858210.984 com.nextcloud.client D/HttpClient: 186546817 start
1598858210.986 com.nextcloud.client D/HttpClient: 11530087 start
1598858210.989 com.nextcloud.client D/HttpClient: 255278525 start
1598858210.991 com.nextcloud.client D/HttpClient: 138550019 start
1598858210.991 com.nextcloud.client D/HttpClient: 53137134 start
1598858210.993 com.nextcloud.client D/HttpClient: 153024697 start
```

I also logged when the operation finished:
```
1598858212.316 com.nextcloud.client D/HttpClient: 236623307 end with size: 2
1598858212.385 com.nextcloud.client D/HttpClient: 203063136 end with size: 2
1598858212.440 com.nextcloud.client D/HttpClient: 123525752 end with size: 2
1598858212.443 com.nextcloud.client D/HttpClient: 220903103 end with size: 2
1598858212.475 com.nextcloud.client D/HttpClient: 240626713 end with size: 2

1598858213.084 com.nextcloud.client D/HttpClient: 205910851 end with size: 2
1598858213.094 com.nextcloud.client D/HttpClient: 237574298 end with size: 2
1598858214.431 com.nextcloud.client D/HttpClient: 30171220 end with size: 2
1598858214.439 com.nextcloud.client D/HttpClient: 115569038 end with size: 2
1598858214.491 com.nextcloud.client D/HttpClient: 17747337 end with size: 2
1598858214.598 com.nextcloud.client D/HttpClient: 240569341 end with size: 2

1598858218.513 com.nextcloud.client D/HttpClient: 96498454 end with size: 2
1598858218.560 com.nextcloud.client D/HttpClient: 241728949 end with size: 2
1598858218.748 com.nextcloud.client D/HttpClient: 16241064 end with size: 2
1598858218.771 com.nextcloud.client D/HttpClient: 22937563 end with size: 2
1598858218.858 com.nextcloud.client D/HttpClient: 54182823 end with size: 2
1598858218.929 com.nextcloud.client D/HttpClient: 82011628 end with size: 2
1598858218.983 com.nextcloud.client D/HttpClient: 139100319 end with size: 2
1598858219.055 com.nextcloud.client D/HttpClient: 19566419 end with size: 2
1598858220.562 com.nextcloud.client D/HttpClient: 13525826 end with size: 2

1598858221.148 com.nextcloud.client D/HttpClient: 260946245 end with size: 2
1598858221.392 com.nextcloud.client D/HttpClient: 263102698 end with size: 2
1598858221.536 com.nextcloud.client D/HttpClient: 122235641 end with size: 2
1598858221.736 com.nextcloud.client D/HttpClient: 46212644 end with size: 2
1598858221.772 com.nextcloud.client D/HttpClient: 66879632 end with size: 2

1598858222.010 com.nextcloud.client D/HttpClient: 233424209 end with size: 2
1598858222.136 com.nextcloud.client D/HttpClient: 190162113 end with size: 2
1598858222.388 com.nextcloud.client D/HttpClient: 249595835 end with size: 2
1598858222.403 com.nextcloud.client D/HttpClient: 145931959 end with size: 2
1598858222.441 com.nextcloud.client D/HttpClient: 166740017 end with size: 2
1598858222.581 com.nextcloud.client D/HttpClient: 77001944 end with size: 2
1598858222.594 com.nextcloud.client D/HttpClient: 181754430 end with size: 2
1598858222.664 com.nextcloud.client D/HttpClient: 162817421 end with size: 2
1598858222.706 com.nextcloud.client D/HttpClient: 53137134 end with size: 2
1598858222.706 com.nextcloud.client D/HttpClient: 22091196 end with size: 2
1598858222.709 com.nextcloud.client D/HttpClient: 222612911 end with size: 2
1598858222.760 com.nextcloud.client D/HttpClient: 186546817 end with size: 2
1598858222.790 com.nextcloud.client D/HttpClient: 166732902 end with size: 2
1598858222.876 com.nextcloud.client D/HttpClient: 138550019 end with size: 2
1598858222.957 com.nextcloud.client D/HttpClient: 130604106 end with size: 2
1598858222.960 com.nextcloud.client D/HttpClient: 12617099 end with size: 2
1598858222.981 com.nextcloud.client D/HttpClient: 148324719 end with size: 2
1598858222.992 com.nextcloud.client D/HttpClient: 254669013 end with size: 2
1598858222.993 com.nextcloud.client D/HttpClient: 255278525 end with size: 2
1598858223.042 com.nextcloud.client D/HttpClient: 45690304 end with size: 2
1598858223.081 com.nextcloud.client D/HttpClient: 153024697 end with size: 2
1598858223.106 com.nextcloud.client D/HttpClient: 264904946 end with size: 2
1598858223.114 com.nextcloud.client D/HttpClient: 72212918 end with size: 2
1598858223.142 com.nextcloud.client D/HttpClient: 83003012 end with size: 2
1598858223.145 com.nextcloud.client D/HttpClient: 11530087 end with size: 2
1598858223.166 com.nextcloud.client D/HttpClient: 130069253 end with size: 2
```

So it seems that they are executed in batches…

**Comment by tobiasKaminsky** [comment]
Some ref:
https://hc.apache.org/httpclient-3.x/performance.html
https://hc.apache.org/httpclient-3.x/threading.html

So it seems that we are doing this all already.

**Comment by kfirfer** [comment]
Hi

I dont know, but its really looks synchronize in the app:

https://files.pica.co.il/s/xRKTGdcoeJPWjeb

**Comment by kfirfer** [comment]
one more screen recording  reference:
https://files.pica.co.il/s/HHeHFsjSZmDSp3W

**Comment by Derkades** [comment]
When doing this, keep in mind to give requests for loading the image a "higher priority". The web interface makes as many requests as it can, with the advantage that it loads very quick (depending on max PHP threads), but the downside is that when opening an image (clicking on a thumbnail) it does not open until thumbnails are loaded for *all* images, which takes a long time for large albums.

**Comment by joshtrichards** [comment]
Fixed in #10080
